### ISSUE TITLE
Bug: tls_connection_policies.client_authentication and http3 produce 403 Forbidden

### ISSUE THREAD

**Issue by xiruizhao** [caddyserver/caddy#4097]
caddy v2.3.0 h1:fnrqJLa3G5vfxcxmOH/+kJOcunPLhSBnjgIvjXV/QTA= on macOS 11.2.3
## Problem:
1. client_authentication off, http3 request => 200
2. client_authentication on, http2 request => 200
3. client_authentication on, http3 request => 403
### configs
client_authentication off, http3 on.
```json
{"apps":{"http":{"servers":{"srv0":{"experimental_http3":true,"listen":[":443"],"logs":{},"routes":[{"handle":[{"body":"hello,world!\n","handler":"static_response"}],"match":[{"host":["localhost"]}],"terminal":true}]}}}},"logging":{"logs":{"default":{"level":"DEBUG"}}}}
```
client_authentication on, http3 on.
with [quic-go/internal/testdata/ca.pem](https://github.com/lucas-clemente/quic-go/blob/master/internal/testdata/ca.pem) in a dummy `tls_connection_policies` **not** matched to current server name (also set a different admin listening port and disable listening on 80)
```json
{"admin":{"listen":":2020"},"apps":{"http":{"servers":{"srv0":{"automatic_https":{"disable_redirects":true},"experimental_http3":true,"listen":[":444"],"logs":{},"routes":[{"handle":[{"body":"hello,world!\n","handler":"static_response"}],"match":[{"host":["localhost"]}],"terminal":true}],"tls_connection_policies":[{"match":{"sni":["localhost"]}},{"client_authentication":{"trusted_ca_certs":["MIICzDCCAbQCCQDA+rLymNnfJzANBgkqhkiG9w0BAQsFADAoMSYwJAYDVQQKDB1xdWljLWdvIENlcnRpZmljYXRlIEF1dGhvcml0eTAeFw0yMDA4MTgwOTIxMzVaFw0zMDA4MTYwOTIxMzVaMCgxJjAkBgNVBAoMHXF1aWMtZ28gQ2VydGlmaWNhdGUgQXV0aG9yaXR5MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA1OcsYrVaSDfhiDppl6oteVspOY3yFb96T9Y/biaGPJAkBO9VGKcqwOUPmUeiWpedRAUB9LE7Srs6qBX4mnl90Icjp8jbIs5cPgIWLkIu8Qm549RghFzB3bn+EmCQSe4cxvyDMN3ndClp3YMXpZgXWgJGiPOylVi/OwHDdWDBorw4hvry+6yDtpQo2TuI2A/xtxXPT7BgsEJDWGffdgZOYXChcFA0c1XVLIYlu2w2JhxS8c2TUF6uSDlmcoONNKVoiNCuu1Z9MorSQmg7a2G7dSPu123KcTcSQFcmJrt+1G81gOBtHB69kacD8xDmgksj09h/ODPL/gIU1ZcU2ci1/QIDAQABMA0GCSqGSIb3DQEBCwUAA4IBAQB0Tb1JbLXp/BvWovSAhO/jwG7UEaUA1rCtkDB+fV2HS9bxCbV5eErdg8AMHKgB51ygUrq95vm/baZmUILr84XKuTEoxxrw5S9Z7SrhtbOpKCumoSeTsCPjDvCcwFExHv4XHFk+CPqZwbMHueVIMT0+nGWss/KecCPdJLdnUgMRz0tIuXzkoRuOiUiZfUeyBNVNbDFSrLigYshTeAPGaYjXCypoHxkeS93nWfOMUu8FTYLYkvGMU5i076zDoFGKJiEtbjSiNW+Hei7u2aSEuCzpqyTKzYPWYffAq3MM2MKJgZdL04e9GEGeuce/qhM1o3q77aI/XJImwEDdut2LDec1"]},"match":{"sni":["not_localhost"]}}]}}}},"logging":{"logs":{"default":{"level":"DEBUG"}}}}
```
### run
checkout v0.20.0 of [quic-go](https://github.com/lucas-clemente/quic-go) and build `example/client`
#### `./client https://localhost:443`
```GET https://localhost:443
client Starting new connection to localhost ([::]:55955 -> 127.0.0.1:443), source connection ID (empty), destination connection ID 82d31f289cf0cda4d84a3a5e40ebfe63706f, version draft-29
Got response for https://localhost:443: &http.Response{Status:"200 OK", StatusCode:200, Proto:"HTTP/3", ProtoMajor:3, ProtoMinor:0, Header:http.Header{"Alt-Svc":[]string{"h3-32=\":443\"; ma=2592000,h3-29=\":443\"; ma=2592000"}, "Server":[]string{"Caddy"}}, Body:(*http3.body)(0xc00041a3f0), ContentLength:-1, TransferEncoding:[]string(nil), Close:false, Uncompressed:false, Trailer:http.Header(nil), Request:(*http.Request)(nil), TLS:(*tls.ConnectionState)(0xc00040a0b0)}
Request Body:
hello, world!
client Closing session with error: Application error 0x100
client Connection 82d31f289cf0cda4d84a3a5e40ebfe63706f closed.
```
Note: `Application error 0x100` means no error.
#### `curl -i --http2 https://localhost:444`
```
HTTP/2 200
alt-svc: h3-32=":444"; ma=2592000,h3-29=":444"; ma=2592000
server: Caddy
content-length: 14
date: Tue, 30 Mar 2021 18:05:33 GMT

hello, world!
```
#### `./client https://localhost:444`
```
GET https://localhost:444
client Starting new connection to localhost ([::]:51384 -> 127.0.0.1:444), source connection ID (empty), destination connection ID d4d9a07e1a7d4cb1ca, version draft-29
Got response for https://localhost:444: &http.Response{Status:"403 Forbidden", StatusCode:403, Proto:"HTTP/3", ProtoMajor:3, ProtoMinor:0, Header:http.Header{"Alt-Svc":[]string{"h3-32=\":444\"; ma=2592000,h3-29=\":444\"; ma=2592000"}, "Server":[]string{"Caddy"}}, Body:(*http3.body)(0xc000120270), ContentLength:-1, TransferEncoding:[]string(nil), Close:false, Uncompressed:false, Trailer:http.Header(nil), Request:(*http.Request)(nil), TLS:(*tls.ConnectionState)(0xc000146000)}
Request Body:

client Closing session with error: Application error 0x100
client Connection d4d9a07e1a7d4cb1ca closed.
```

**Comment by francislavoie** [comment]
FYI @marten-seemann

**Comment by francislavoie** [comment]
I only spotted this issue again just now when reading through old issues, and I've since realized that the `403` comes from `strict_sni_host` being implicitly enabled when client authentication is enabled.

While this is the older issue, #4294 has seen more discussion since so I'll close this one in favor of that one. If you have any help to offer in testing to figure out why that case is triggering, that would help. I don't have an environment set up in which I can properly test HTTP3 at the moment.
